flareon/src/lib.rs

```diff
@@ -1,14 +1,305 @@
-pub fn add(left: u64, right: u64) -> u64 {
-    left + right
+pub mod prelude;
+
+use std::fmt::{Debug, Formatter};
+use std::io::Read;
+use std::sync::Arc;
+
+use async_trait::async_trait;
+use axum::handler::HandlerWithoutStateExt;
+use bytes::Bytes;
+use derive_builder::Builder;
+use indexmap::IndexMap;
+use log::info;
+use thiserror::Error;
+
+pub type StatusCode = axum::http::StatusCode;
+
+#[async_trait]
+pub trait RequestHandler {
+    async fn handle(&self, request: Request) -> Result<Response, Error>;
+}
+
+#[derive(Clone, Debug)]
+pub struct Router {
+    urls: Vec<Route>,
+}
+
+impl Router {
+    #[must_use]
+    pub fn with_urls<T: Into<Vec<Route>>>(urls: T) -> Self {
+        Self { urls: urls.into() }
+    }
+
+    async fn route(&self, request: Request, request_path: &str) -> Result<Response, Error> {
+        for route in &self.urls {
+            if request_path.starts_with(&route.url) {
+                let request_path = &request_path[route.url.len()..];
+                match &route.view {
+                    RouteInner::Handler(handler) => return handler.handle(request).await,
+                    RouteInner::Router(router) => {
+                        return Box::pin(router.route(request, request_path)).await
+                    }
+                }
+            }
+        }
+
+        unimplemented!("404 handler is not implemented yet")
+    }
+}
+
+#[async_trait]
+impl RequestHandler for Router {
+    async fn handle(&self, request: Request) -> Result<Response, Error> {
+        let path = request.uri().path().to_owned();
+        self.route(request, &path).await
+    }
+}
+
+#[async_trait]
+impl<T> RequestHandler for T
+where
+    T: Fn(Request) -> Result<Response, Error> + Send + Sync,
+{
+    async fn handle(&self, request: Request) -> Result<Response, Error> {
+        self(request)
+    }
+}
+
+/// A building block for a Flareon project.
+///
+/// A Flareon app is a part (ideally, reusable) of a Flareon project that is
+/// responsible for its own set of functionalities. Examples of apps could be:
+/// * admin panel
+/// * user authentication
+/// * blog
+/// * message board
+/// * session management
+/// * etc.
+///
+/// Each app can have its own set of URLs that it can handle which can be
+/// mounted on the project's router, its own set of middleware, database
+/// migrations (which can depend on other apps), etc.
+#[derive(Clone, Debug, Builder)]
+#[builder(setter(into))]
+pub struct FlareonApp {
+    router: Router,
+}
+
+impl FlareonApp {
+    #[must_use]
+    pub fn builder() -> FlareonAppBuilder {
+        FlareonAppBuilder::default()
+    }
+}
+
+impl FlareonAppBuilder {
+    #[allow(unused_mut)]
+    pub fn urls<T: Into<Vec<Route>>>(&mut self, urls: T) -> &mut Self {
+        let mut new = self;
+        new.router = Some(Router::with_urls(urls.into()));
+        new
```

Comment:
Out of curiosity, why the indirection?

Reply:
Ah, honestly just mindlessly copied the output from the derive-builder crate. No, I don't think there's any good reason to introduce the `new` variable. Removed.